Removing anonymous variables assignments from output substitution

traiansf opened the pull request:
Addressing Issue #254

traiansf committed:
Removing anonymous variables assignments from output substitution

dwightguth reviewed (changes requested):
So, K already has code to prune substitutions of variables that are tagged with the anonymous attribute in the k frontend. You should be adjusting that code if you need to rather than reinventing the wheel.

traiansf commented:
@dwightguth could you point to a location?  I wouldn't want to convert this into a JavaBackend substitution just to be able to call that pruning code.  The code I wrote would work for any backend outputing substitutions in Kore format.

dwightguth reviewed (changes requested):
it looks like you're correct that that was being handled inside the java backend previously. still, I have a couple requests.

dwightguth commented on haskell-backend/src/main/java/org/kframework/backend/haskell/HaskellRewriter.java:
```diff
@@ -245,7 +248,8 @@ public K search(K initialConfiguration, Optional<Integer> depth, Optional<Intege
                         K outputK = koreToK.apply(kore);
-                        return outputK;
+                        K trimmedOutput = SubstitutionPruning.pruneSubstitution(generatedVars, outputK);
+                        return trimmedOutput;
```
I would like you to return the original output here but with the "anonymous" attribute added to the correct anonymous variables. And then move the substitution trimming code to KPrint.java and run it only when we are pretty printing.

traiansf replied:
makes sense

dwightguth commented on kore/src/main/scala/org/kframework/compile/SubstitutionPruning.scala:
```diff
@@ -0,0 +1,55 @@
+package org.kframework.compile
```
I would like you to rewrite this code in Java. We are trying to reduce the amount of scala code we have in the code base, not increase it.

traiansf replied:
Got it!  It wasn't clear to me so far (since my work was mostly on scala code lately) ...

dwightguth reviewed (changes requested):
it looks like you are correct that this was happening inside the java backend previously. Still, I would like you to fix a few things.

traiansf commented:
I've made the following changes:

- translated the removal code to java
- added the call to the removal code in `KRunExecutionMode`
- removed the code from the Java backend that achieved the same thing for the Java backend
- modified java backend serialization of search results to use ML `#Equals` instead of `==K`
- added a krun print option to control whether assignments of anonymous variables are printed

Still to do:
- Update the tests which are failing because of the change from `==K` to `#Equals`

@dwightguth could you please take a look to check whether you are OK with the changes so far?  If so, I'll make another commit with the tests changes.

dwightguth reviewed (changes requested):
This looks fine, but term normalization happens during pretty printing in KPrint.java, not in KRunExecutionMode. You should revert those changes and move it to the right place, otherwise people who request the raw term from k will get something different than we want. Look for the code relating to sorting collections and alpha renaming, that's where this goes.

traiansf commented:
Well, I know that you proposed to put it in `KPrint`, but I don't think that it should be there because  by the time a pattern reaches KPrint it is simply a term, information on where it comes from is lost;  the problem I'm trying to address is specific to the output of a pattern match (which is supposed to have a specific form) and I'm not convinced that it should be applied on all potential K terms.

If you instead insist it should be there, I'm not sure how to ensure that for any K term, given in any form, once can (soundly) eliminate all equalities in which the LHS is an anonymous variable.  Do we even want to eliminate all of them?

dwightguth commented:
I mean it seems to me that any output term that has the form of a substitution should have this change applied to it.

traiansf committed:
modified tests

traiansf commented:
Jenkins: test this please

traiansf commented:
Jenkins: test this please

ehildenb commented:
Can one of the admins verify this patch?

traiansf commented:
@dwightguth I've tested it on a separate machine and confirmed that the changes in this PR make krun seemingly loop forever for certain tests, so I'm going to have to debug it further.

traiansf committed:
Fixed mini.KApply hashCode bug && added toString methods

traiansf commented:
@dwightguth can you please re-review?  Apparently the looping behavior was caused by a bug in the hashcode implementation of mini.KApply which was triggering an exponential (in number of KApply nodes) evaluation of hashcodes.

dwightguth commented:
Looks okay but I still need to review the tests which I can't do on GitHub. I'll try to get to this later today.

dwightguth commented:
I'm closing this as it has bitrotted. If you still need this functionality, feel free to fix the conflicts and issues and make a new PR.
